### DISCUSSION TITLE
GitHub breaking casks with main vs master

### DISCUSSION THREAD

**Discussion by guruevi** [Homebrew/discussions#592] [Casks]
Apparently some idiots decided renaming Git branches without referencing is a good thing. This has been racking me all morning, a ton of my scripts to deploy new machines no longer work properly. I can't fix it as far as I can see as the particular git command is being generated by Brew and I don't know how to override it. I've tried a variety of other things for example downloading the caches myself, but then the pull request still looks for the master branch.

Some branches from major companies (Microsoft, PuppetLabs etc) are now called 'main' instead of 'master' breaking pretty much everything in the CI world. Brew seems to automatically fetch 'master', which works for some repos, but not for others.

This isn't a bug per-se in the Brew system, but it is a problem.

Command that failed
```
brew tap kmalinich/facter
brew install --head facter

Failure while executing; `git clone --depth 1 --branch master -c advice.detachedHead=false https://github.com/puppetlabs/leatherman.git /Users/isdadmin/Library/Caches/Homebrew/l
eatherman--git` exited with 128.
Cloning into '/Users/isdadmin/Library/Caches/Homebrew/leatherman--git'...                                                                                                        
warning: Could not find remote branch master to clone.                                                                                                                           
fatal: Remote branch master not found in upstream origin    
Output of command with --verbose --debug
```

I tried manually cloning into the above location to no avail

**Comment by gromgit** [comment]
Did you run `brew install` _immediately_ after `brew tap`, or did you tap it some time ago (and therefore the branch renaming happened _after_ the tap)?

If the former, something's wrong indeed, because the `git clone` that's the core of `brew tap` should've picked up the current branch info.

If it's the latter, you can simply:
```
brew untap kmalinich/facter
brew tap kmalinich/facter
```
to reset git's (not Homebrew's) understanding of the tap's new branches.

**Reply by guruevi** [reply]
Nope, this is a brand new installation of macOS + Brew. Tested both on MacBook Pro M1 with Big Sur and Intel machine with Catalina to make sure. 

The git command is injected by Brew. The tap repo doesn't even reference a branch (https://github.com/kmalinich/homebrew-facter).

**Reply by guruevi** [reply]
Basically, the issue is, per the HomeBrew documentation here: 

https://rubydoc.brew.sh/Formula.html#head-class_method
The download strategies (e.g. :using =>) are the same as for url. master is the default branch and doesn’t need stating with a branch: parameter.

This assumption, which has been true since the inception of Git and still is the standard according to the Git documentation, is no longer true on GitHub. GitHub no longer creates and is actively complicit in the cloning and then deleting of the master branches which breaks decades of established process and Git.

The question is whether we now have to create tons of shadow-github branches that 'fix' these things to comply with the latest demands from Microsoft or is it feasible to use another public git hosting service that doesn't break things.

**Reply by gromgit** [reply]
Sorry, misread the error. Yeah, Homebrew assumes `master` branches when downloading from Git repos during the build process.

@kmalinich would have to change:
```
  head "https://github.com/puppetlabs/leatherman.git"
```
to:
```
  head "https://github.com/puppetlabs/leatherman.git", branch: "main"
```
in `leatherman.rb`. Try raising an issue at that repo, then it gets fixed for all users of that tap.

**Comment by kmalinich** [comment]
Holy crap, somebody is using the tap I made?
Sorry for the delay, I just approved and merged the PR
